### ISSUE TITLE
Signature verification, no error message

### ISSUE THREAD

**Issue by LionelJouin** [SEP-G5/Backend#65]
I'm working on a fix for this issue https://github.com/SEP-G5/Mobile-Client/issues/39 
During the verification of the signature and the transaction ([transaction.rs#L148](https://github.com/SEP-G5/Backend/blob/master/src/blockchain/transaction.rs#L148)), the 'match' send me in the Err without any error message, only empty parentheses

I modified a bit the code to print the error line 157
```
                Err(e) => {
                        println!("error : {:?}", e);
```

**Comment by dumheter** [comment]
That doesn't sound very helpful. How can I reproduce this empty error, could you provide the content of the REST call?

**Comment by LionelJouin** [comment]
You can reproduce it with this body
```
{
	"id": "abc123",
	"timestamp": 1577123330,
	"publicKeyOutput": "_tzBYJVICsit2iJPYeAldD9_QrLi5BuEsI3eNbt8YFk",
	"signature": "8D-xgsjncmPfiY4y8enEoRMhw4QlpdMbsAkoOpL9VuhDlcQXFj5X_HU4RRVbcR2o-gMwa40JirXLb-BWjbwsCQ"
}
```

The private key is
```
JgABDE8ISmaeJ86VvmE1Qn2DAryanATeQ2fkK6vEqC4
```

**Comment by Alfret** [comment]
libsodium does not give us any reason for the failure, so we sadly cannot give more information. However the format of the public key does not seem right (there is always a '=' at the end), which can be caused by padding or something. The signature is also not correct, way too short.

A valid transaction may look like this:

Valid transaction:
```json
{
  "id": "Björnberg_634",
  "publicKeyInput": null,
  "publicKeyOutput": "T4ciUOjQ5yldcWSPV1OoKyJzikWLOHVubat5UNyUhBg=",
  "signature": "b7Rs-uBNWVh5Uu5-2aMBIiyOPgImwIEJsIccbrL_Nk0rEjthXJUkuWKRiA7BdihtNbNgE_uIKpQk8Bfr6-gXA0Jqw7ZybmJlcmdfNjM0VG0UXgAAAABPhyJQ6NDnKV1xZI9XU6grInOKRYs4dW5tq3lQ3JSEGA==",
  "timestamp": 1578397012
}
```

Another valid transaction:
```json
{
  "id": "Henriksson_806",
  "publicKeyInput": null,
  "publicKeyOutput": "Xky_WxVuWfDY0WF3cUg3v6GgCDjH_n2Mg_79GtxBgAc=",
  "signature": "h_GI_RPK9Whd2WyGFB2wrRtjOPmM6HxkXETDVmrbCFX3SQoVrvsRCmQuwl-qauXShdi6vsVxHQD8vjOYF9z3CEhlbnJpa3Nzb25fODA2WW0UXgAAAABeTL9bFW5Z8NjRYXdxSDe_oaAIOMf-fYyD_v0a3EGABw==",
  "timestamp": 1578397017
}
```
The public key and signature are encoded as base64

**Comment by LionelJouin** [comment]
I'm using elliptic which provides eddsa ed25519 key pair generation, I get a 32 bytes array I transform to a base64 url safe string, I don't get any error from the backend about it. I can encrypt, decrypt, sign and verify with these keys. 
To sign the data, I hash the object using sha256, then I encode the signature in base64 url safe.

**Comment by dumheter** [comment]
@LionelJouin could you try print the raw bytes of the public key you provided? This is what the byte64 encoded text you provided gets translated to:
```
[254, 220, 193, 96, 149, 72, 10, 200, 
 173, 218, 34, 79, 97, 224, 37, 116, 
 63, 127, 66, 178, 226, 228, 27, 132, 
 176, 141, 222, 53, 187, 124, 96, 89]
```
If they match that would show that the base64 stuff is correct, and the problem lies elsewhere.

**Comment by LionelJouin** [comment]
This is the raw bytes of the key when I print it:
```
[254, 220, 193, 96, 149, 72, 10, 200, 173, 218, 34, 79, 97, 224, 37, 116, 63, 127, 66, 178, 226, 228, 27, 132, 176, 141, 222, 53, 187, 124, 96, 89]
```

**Comment by dumheter** [comment]
Okey, good. It is not a base64 issue. My next thought of problem area is the signature. How do you make your signature? This is how its done in the rust code:
```rust
pub(crate) fn sign(&mut self, sk: &SecretKey) {
        let buf = self.content_to_u8();
        let sig = sign(buf.as_slice(), &sk);
        self.signature = sig;
    }
```
We take the entire content of the transaction and produce a signature from it. I'm thinking having public_key_input be none might impact the rust code and javascript code differently?.

**Comment by dumheter** [comment]
```javascript
static sign(privateKey, data) {
        return new Promise(async function (resolve) {
            const ec = new eddsa('ed25519');
            privateKey = toBuffer(privateKey);
            const key = ec.keyFromSecret(privateKey);
            const hash = Cryptography.getDataHash(data);
            const signature = encode(key.sign(hash).toBytes());
            resolve(signature);
        });
```
It looks like your are producing your signature differentely, signing a hash of the data instead. Why are you doing it?

**Comment by LionelJouin** [comment]
To sign the data, I hash the object using sha256, then I sign it, the public_key_input is not included if it is null.

**Comment by dumheter** [comment]
Why are you hashing the data before signing it, should we do it too?

**Comment by LionelJouin** [comment]
I was thinking you hashed the data before signing it, I have to find where I read it. Even if I remove the hashing, it doesn't work.

**Comment by dumheter** [comment]
I think hashing the data before signing it is a good idea, it limits the length of the signature. We will do this in issue #69 . 

I think the last problem is just a difference in how we are signing.
```rust
fn content_to_u8(&self) -> Vec<u8> {
        let mut buf: Vec<u8> = Vec::from(self.id.as_bytes());
        buf.extend_from_slice(&self.timestamp.to_le_bytes());
        if let Some(ref key) = self.pub_key_input {
            buf.extend(key);
        }
        buf.extend(&self.pub_key_output);
        buf
    }
```
This function is called on a transaction. It takes all data and puts them in a u8 buffer. The order of each element is important. Are we doing it the same way?

**Comment by LionelJouin** [comment]
Yes, my data were not in a u8 buffer, I fixed it using Uint8Array in javascript, but now I get the error "content does not match the signature". 
The order of my data is: id - timestamp - input key (if exists) - output key

**Comment by dumheter** [comment]
@LionelJouin I wrote to you on slack, lets talk about it on discord or meet up, since we are not able to solve it here.

**Comment by dumheter** [comment]
Fixed in commit c69f4d729bfe3f9f6a026c0be591087744c2d8c6
